e3db/identity.py

```diff
@@ -6,17 +6,94 @@
-    TODO:  Contains a realm, have a client, have all of its OAuth tokens
+    TODO WASH-902:  Contains a realm, a storage client, have all of its OAuth tokens
     """
+    DEFAULT_API_URL = "https://api.e3db.com"
 
+    def __init__(self):
+        """
+        TODO
+        """
 
     @staticmethod
-    def derive_note_creds(user_name: str, password: str, realm_name: str):
+    def identity_login(user_name: str, password: str, realm_name, app_name: str, api_url: str=DEFAULT_API_URL) -> dict:
+        """
+        A factory mehod to login an existing user, get the stored identity credentials for a user,
+        create a client for them, and return an Identity object.
+
+        TODO: Change param from realm_name & app_name to a Realm Object
+        TODO WASH-902: Currently returns a config object - need to create an Id_Client class to create with the config
+
+        The username and password are used to derive encryption keys used to fetch a pre-stored
+        note which contains the users identity credentials.
+
+        Only password based realm types are supported at this time.
+        Brokered Login is not supported at this time.
+        MFA is not supported
+
+        Parameters
+        ----------
+        user_name : str
+
+        password : str
+
+        realm_name : str
+            will be converted to lower case
+        
+        app_name : str
+
+        api_url: str
+            Defaults to https://api.e3db.com
+
+        Retruns
+        -------
+        dict
+            config object for creating an identity client
+        """
+        realm_name = realm_name.lower()
+        note_name, key_pair, signing_key_pair = Identity.derive_note_creds(user_name, password, realm_name)
+        pkce_verifier, pkce_challenge = Crypto.generate_pkce_challenge()
+        auth = E3DBTSV1Auth(signing_key_pair.private_key)
+
+        # PKCE step #1: submit a challenge for this user & Realm
+        redirect = pkce_submit_challenge(user_name, realm_name, app_name, pkce_challenge, auth, api_url)
+        
+        #PKCE step #2, submit public keys to redirect with session params
+        context = pkce_submit_keys(key_pair, signing_key_pair, redirect, auth)    
+
+        # PKCE step #3 make final request for creds using context and verifier
+        credentials = pkce_get_credentials(realm_name, context, pkce_verifier, auth, api_url)
+        access_token = credentials["access_token"]
+
+        # Read Id note using our access token 
+        stored_creds = Client.read_anonymous_note_by_name(note_name, 
+                                                    key_pair.private_key,
+                                                    signing_key_pair.private_key,
+                                                    auth_headers={ TOZID_LOGIN_HEADER : access_token },
+                                                    api_url=api_url)
+
+        # TODO: add backwards compatibility consistent with JS-SDK for old config format
```

Comment:
This todo can be removed, we won't port that code over

Reply:
Whats the user impact of this TODO not being done?

Reply:
This was a super early interface change that we made and the JS-SDK has a migration path if it detects the old style. Our understanding is that no existing users have the old-style notes and we don't need to propagate the code.

Reply:
removed